src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/MouseInputActionObserver.cs

```diff
@@ -207,18 +224,63 @@ public MouseInputActionData GetCurrentMouseState()
 
         private readonly ConcurrentQueue<MouseInputActionData> _mouseInputActions = new();
 
-        public List<MouseInputActionData> FlushInputDataBuffer(float upToTime = float.MaxValue)
+        public void ClearBuffer()
+        {
+            _mouseInputActions.Clear();
+        }
+
+        public List<MouseInputActionData> FlushInputDataBuffer(bool ensureOne = false)
         {
             List<MouseInputActionData> result = new();
             while (_mouseInputActions.TryPeek(out var action))
             {
-                if (action.startTime < upToTime)
+                _mouseInputActions.TryDequeue(out _);
+                result.Add(action);
+            }
+
+            if (ensureOne && result.Count == 0 && _priorMouseState != null)
+            {
+                // or.. we need at least 1 mouse observation per tick, otherwise hover over effects/etc don't function correctly
+                result.Add(_priorMouseState);
+            }
+
+            return result;
+        }
+
+        private IEnumerable<RecordedGameObjectState> FindObjectsAtPosition(Vector2 position, List<RecordedGameObjectState> state, out float maxZDepth)
+        {
+            // make sure screen space position Z is around 0
+            var vec3Position = new Vector3(position.x, position.y, 0);
```

Comment:
If for a given game this is not the case, is that ok?

Reply:
this has nothing to do with any specific game..

when projecting world space to screen space, the numbers always come with a Z offset from the camera

for simplicity of comparison to mouse clicks, I exclude that information from what is recorded and handle the max zOffset as a separate field on the object.  This allows us to do Bounds.contains for mouse positions without it erroneously returning false

Reply:
Ah ok, I misread the comment, that makes more sense to me now